### PR TITLE
Fix #146 - More optimized reduction functions

### LINKED ISSUE #146: Optimized reduction
Use shuffle to reimplement all_reduce and subsequent numerical reduction
Require #145 
- [x] all_reduce
- [x] splatted modifier
- [x] autofold / all_reduce combo
- [x] sum,prod,min,max

### PR DESCRIPTION
**Pull request by jfalcou**
- Proper set-up for all reduction like sum, prod, minimum,maximum, the boolean reduction
- Support for the splatted_ decorator
- Optimizing sum for various specific x86 variants

### CODE REVIEW

File: include/boost/simd/arch/x86/sse2/simd/function/sum.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  BOOST_DISPATCH_OVERLOAD ( sum_
+                          , (typename A0)
+                          , bs::sse2_
+                          , bs::pack_<bd::ints16_<A0>, bs::sse_>
+                         )
+  {
+    BOOST_FORCEINLINE typename A0::value_type operator()(const A0& a0) const BOOST_NOEXCEPT
+    {
+      // no int16 shuffle, so we optimize using other ways
```
Comment (mgaunard) [inline include/boost/simd/arch/x86/sse2/simd/function/sum.hpp:84]: couldn't this be handled by autofold or whatever it's called?
If there is no native shuffle allowing to do log2 reduction, then do a normal linear one.

Comment (jfalcou) [inline include/boost/simd/arch/x86/sse2/simd/function/sum.hpp:84]: autofold use the butterfly recursive shuffle pattern. In some cases, includign this one, the full shuffle is some cycles slower than a very specific implementation. It's mostly true for X86 as most other SIMD ISA don't have so much speicalized copy/shuffle/half-sum stuff.

In the end, it's a matter of latency and so on. This implementation is roughly 40% faster than the old one we had before.

Comment (mgaunard) [inline include/boost/simd/arch/x86/sse2/simd/function/sum.hpp:84]: I understand that, which is why I said that you could detect whether there is native shuffle or not and do this implementation instead.
This fold is still generic, it's not specific to sum.

Comment (jfalcou) [inline include/boost/simd/arch/x86/sse2/simd/function/sum.hpp:84]: autofold is used for all others fold (prod, etc). sum has some specific optimizations here which I don't feel like overgeneralizing to remove this one case. You have a point on something else it that detecting performant native shuffle is somethign to have somehow for other cases. I'll see how it turns out.

**Pull request merged**